feat: support filtering users table by login type

utsavll0 opened the pull request:
#15896 Mentions ability to add support for filtering by login type

The issue mentions that backend API support exists but the backend did not seem to have the support for this filter. So I have added the ability to filter it.

I also added a corresponding update to readme file to make sure the docs will correctly showcase this feature

utsavll0 committed:
ability to filter users based on login type

utsavll0 committed:
update documentation also

utsavll0 commented:
I have read the CLA Document and I hereby sign the CLA

ethanndickson commented:
Thanks for the PR!

Looking good so far, but we'll need some tests. I believe the right place for those is `coderd/users_test.go`

utsavll0 committed:
add tests

utsavll0 committed:
Merge branch 'main' into main

utsavll0 commented:
@ethanndickson I added a few tests for the PR and I also ran the commands which you mentioned

utsavll0 committed:
change test case

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1904,6 +1904,43 @@ func TestGetUsers(t *testing.T) {
+
+	// Create a user with a specific role
+	_, err := client.User(ctx, first.UserID.String())
+	require.NoError(t, err, "")
```
```suggestion
	require.NoError(t, err)
```

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1904,6 +1904,43 @@ func TestGetUsers(t *testing.T) {
 	})
 }
 
+func TestGetUsersFilters(t *testing.T) {
```
We already have a test called `TestGetUser`, and we already have a test called `TestUsersFilter`, (they're both very similar, unfortunately). Since your test has a narrower scope, it should be named a bit more appropriately. Alternatively, you could add extra `t.Run` cases to either existing test. I would personally just add extra cases to `TestGetUsers` here, but it's up to you.

ethanndickson reviewed (commented):
Your newly written tests are failing as, by default, DB queries are ran against a fake DB in `dbmem.go`. You'll need to update the `GetUsers` Go function with the same logic as the Postgres query. 

To run `go test` against postgres, just set `DB=ci`: 
`DB=ci go test -run "TestGetUsersFilters"  github.com/coder/coder/v2/coderd`
(This is essentially what `make test-postgres` does)

utsavll0 committed:
filter by multiple login types

utsavll0 committed:
add filter by logintype to mock db

utsavll0 committed:
update tests

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		t.Parallel()
+		client := coderdtest.New(t, nil)
+		first := coderdtest.CreateFirstUser(t, client)
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
```
```suggestion
		ctx := testutil.Context(t, testutil.WaitLong)
```
I see none of the other tests in this file use this util, but we might as well use it here.

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		t.Parallel()
+		client := coderdtest.New(t, nil)
+		first := coderdtest.CreateFirstUser(t, client)
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
```
```suggestion
		ctx := testutil.Context(t, testutil.WaitLong)
```

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		t.Parallel()
+		client := coderdtest.New(t, nil)
+		first := coderdtest.CreateFirstUser(t, client)
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
```
```suggestion
		ctx := testutil.Context(t, testutil.WaitLong)
```

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		res, err := client.Users(ctx, codersdk.UsersRequest{
+			LoginType: []codersdk.LoginType{codersdk.LoginTypeNone},
+		})
+		require.NoError(t, err, "should not error when filtering by role")
```
```suggestion
		require.NoError(t, err)
```
I wouldn't bother justifying a requirement as trivial as no error, unless you think it's unclear why no error is expected.

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		})
+		require.NoError(t, err)
+
+		// Test filtering by role
```
```suggestion
```

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+			LoginType: []codersdk.LoginType{codersdk.LoginTypeNone},
+		})
+		require.NoError(t, err, "should not error when filtering by role")
+		require.Len(t, res.Users, 1, "should find one user with the member role")
+		require.Equal(t, res.Users[0].Username, "bob", "should return the correct user with member role")
```
I'm not sure the user having the member role is relevant here? The test is simple enough that not having a comment for either of these is fine.

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		})
+		require.NoError(t, err)
+
+		// Test filtering by role
```
```suggestion
```

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		res, err := client.Users(ctx, codersdk.UsersRequest{
+			LoginType: []codersdk.LoginType{codersdk.LoginTypeNone, codersdk.LoginTypeGithub}, // Ensure we're filtering by login type
+		})
+		require.NoError(t, err, "should not error when filtering by role")
```
```suggestion
		require.NoError(t, err)
```
The query isn't filtering by role.

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		})
+		require.NoError(t, err, "should not error when filtering by role")
+		require.Len(t, res.Users, 2, "should find two users with the specified login types")
+		usernames := make(map[string]bool)
+		for _, user := range res.Users {
+			usernames[user.Username] = true
+		}
+		require.True(t, usernames["bob"], "should return the correct user with login type none")
+		require.True(t, usernames["charlie"], "should return the correct user with login type github")
+		// Ensure that the user with login type none is indeed in the result
+		for _, user := range res.Users {
+			if user.Username == "bob" {
+				require.Equal(t, user.LoginType, codersdk.LoginTypeNone, "bob should have login type none")
+			}
+			if user.Username == "charlie" {
+				require.Equal(t, user.LoginType, codersdk.LoginTypeGithub, "charlie should have login type github")
+			}
+		}
+	})
```
This is kind of hard to follow. You could copy what some of the other tests do, and store the return value of each `CreateUserWithOrgs`, and then compare that against the output of `client.Users`, using `ElementsMatch`.

ethanndickson commented on coderd/users_test.go:
```diff
@@ -1902,6 +1902,106 @@ func TestGetUsers(t *testing.T) {
+		require.NoError(t, err, "should not error when filtering by role")
+		require.Len(t, res.Users, 1, "should find one dormant user with the specified login type")
+		require.Equal(t, res.Users[0].Username, "bob", "should return the correct dormant user with login type none")
+		require.Equal(t, res.Users[0].Status, codersdk.UserStatusSuspended, "bob should be in dormant status")
```
I think a lot of these comments are just restating what's already implicit in the code, and aren't useful. We already know `res.Users[0]` is bob from the line above.

ethanndickson reviewed (commented):
Nice work on these tests! Just a few comments, mostly on style.

CI is all passing except for `weekly-docs / check-docs`, which you can safely ignore (it should go away if you merge / rebase your branch off main)

utsavll0 committed:
cleanup tests

utsavll0 committed:
Merge branch 'main' into main

utsavll0 commented:
Thank you for your feed @ethanndickson. I am to blame for these silly errors because of my over-reliance on copilot. I have fixed the style and removed unnecessary comments from the code. Please have a look again.

ethanndickson commented on coderd/users_test.go:
These tests are looking good, but there's a bit of a blindspot: all the expected outputs are equal to all the users created in the test. It would be good to have some tests that ensure users that don't match the query get excluded.

ethanndickson replied:
That's a good point too, one more test on excluding users explicitly would be good enough I reckon.

ethanndickson commented on codersdk/users.go:
```diff
@@ -750,6 +751,11 @@ func (c *Client) Users(ctx context.Context, req UsersRequest) (GetUsersResponse,
 			if req.SearchQuery != "" {
 				params = append(params, req.SearchQuery)
 			}
+			if len(req.LoginType) > 0 {
```
The length check here is redundant, since it's immediately followed by a for-range loop

utsavll0 committed:
remove redundant len check

utsavll0 committed:
add test to filter 1 user among many

utsavll0 commented:
ok @ethanndickson I think this final test should do it

utsavll0 committed:
Merge branch 'main' into main

ethanndickson commented on docs/admin/users/index.md:
```diff
@@ -190,6 +190,8 @@ to use the Coder's filter query:
   `status:active last_seen_before:"2023-07-01T00:00:00Z"`
 - To find users who were created between January 1 and January 18, 2023:
   `created_before:"2023-01-18T00:00:00Z" created_after:"2023-01-01T23:59:59Z"`
+- To find users who have login type as github and is a member:
+    `login_type:github role:member`
```
```suggestion
- To find users who log in using GitHub:
    `login_type:github`
```
`role:member` shows everyone, as everyone is a member.

ethanndickson commented on docs/admin/users/index.md:
```diff
@@ -203,3 +205,4 @@ The following filters are supported:
   the RFC3339Nano format.
 - `created_before` and `created_after` - The time a user was created. Uses the
   RFC3339Nano format.
+- `login_type` - Represents the login type of the user. Refer here for all the roles [LoginType documentation](https://pkg.go.dev/github.com/coder/coder/v2/codersdk#LoginType)
```
```suggestion
- `login_type` - Represents the login type of the user. Refer to the [LoginType documentation](https://pkg.go.dev/github.com/coder/coder/v2/codersdk#LoginType) for a list of supported values.
```

ethanndickson reviewed (commented):
Looking good, just two comments on the copy.

utsavll0 committed:
doc update

utsavll0 committed:
capitalize correctly

ethanndickson reviewed (dismissed):
Thanks a lot for the PR and your patience through these review cycles! ❤️

utsavll0 commented:
Thank you @ethanndickson for staying with me. This was my first ever PR in Go and I learned a lot. 🥳 🍾
